tasks/run_agent.rb

```diff
@@ -0,0 +1,160 @@
+#!/opt/puppetlabs/puppet/bin/ruby
+
+require 'open3'
+require 'etc'
+require 'json'
+require 'yaml'
+require_relative '../../ruby_task_helper/files/task_helper.rb'
+
+class RunAgentTask < TaskHelper
+  def task(puppet_bin: '/opt/puppetlabs/bin/puppet', wait_time: 300, _noop: false, **_kwargs)
+    @puppet_bin = puppet_bin
+    @wait_time = wait_time
+    @noop = _noop
+
+    check_running_as_root
+    check_agent_not_disabled
+    wait_for_puppet_lockfile
+    run_puppet
+  end
+
+  def wait_for_puppet_lockfile
+    waited = 0
+    while agent_locked? && waited < @wait_time
+      sleep 1
+      waited += 1
+    end
+
+    if agent_locked?
+      raise TaskHelper::Error.new(
+        'Lockfile still exists after waiting',
+        'theforeman-puppet/lockfile_timeout_expired'
+      )
+    end
+
+    waited
+  end
+
+  def check_agent_not_disabled
+    if agent_disabled?
+      raise TaskHelper::Error.new(
+        'Agent is disabled on this node',
+        'theforeman-puppet/agent_disabled',
+        disabled_reason
+      )
+    end
+  end
+
+  def agent_locked?
+    File.exist?(agent_catalog_run_lockfile)
+  end
+
+  def agent_disabled?
+    File.exist?(agent_disabled_lockfile)
+  end
+
+  def disabled_reason
+    file = File.open agent_disabled_lockfile
+    data = JSON.load file
+    file.close
+    data
+  end
+
+  def agent_disabled_lockfile
+    return @agent_disabled_lockfile unless @agent_disabled_lockfile.nil?
+
+    @agent_disabled_lockfile = agent_config('agent_disabled_lockfile')
+    @agent_disabled_lockfile
+  end
+
+  def agent_catalog_run_lockfile
+    return @lockfile unless @lockfile.nil?
+
+    @lockfile = agent_config('agent_catalog_run_lockfile')
+    @lockfile
+  end
+
+  def lastrunfile
+    return @lastrunfile unless @lastrunfile.nil?
+
+    @lastrunfile = agent_config('lastrunfile')
+    @lastrunfile
+  end
+
+  def agent_config(setting)
+    cmd = [@puppet_bin, 'config', 'print', setting]
+    stdout, stderr, status = Open3.capture3(*cmd)
+
+    unless status.exitstatus.zero?
+      raise TaskHelper::Error.new(
+        "Couldn't determine #{setting} configuration",
+        'theforeman-puppet/config_unknown',
+        stderr: stderr
+      )
+    end
+    stdout.strip
+  end
+
+  def check_running_as_root
+    unless Process.euid.zero?
+      raise TaskHelper::Error.new(
+        'Puppet agent needs to run as root',
+        'theforeman-puppet/bad_euid',
+        euid: Process.euid
+      )
+    end
+  end
+
+  def run_puppet
+    cmd = [@puppet_bin, 'agent', '-t', '--detailed-exitcodes']
```

Comment:
Sorry, with https://github.com/theforeman/puppet-puppet/pull/709#issuecomment-544475964 I was of course referring to this line.

Reply:
For a manual run (as opposed to a normal scheduled run), I think I'd personally never want to accidentally be using the cache.  I do need a way of making this and other options configurable though.

Reply:
`--detailed-exitcodes` is redundant here though (already included in `-t`).

Reply:
Yes, agreed, having a simple toggle for `--test` (which doesn't use cache) and `--no-daemonize --onetime --no-splay --detailed-exitcodes` (which would use cache) would be nice.

Reply:
Can you override options set with `--test` by setting conflicting arguments?
eg does
```
puppet agent -t --usecacheonfailure
```
work?

Reply:
answering my own question... Apparently not.

Reply:
Seems like it:

~~~
$ puppet apply --noop -e 'file { "/tmp/omgtest": ensure => file, }'
Notice: Compiled catalog for managua.lan in environment production in 0.01 seconds
Notice: /Stage[main]/Main/File[/tmp/omgtest]/ensure: current_value 'absent', should be 'file' (noop)
Notice: Class[Main]: Would have triggered 'refresh' from 1 event
Notice: Stage[main]: Would have triggered 'refresh' from 1 event
Notice: Applied catalog in 0.02 seconds

$ puppet apply --noop -e 'file { "/tmp/omgtest": ensure => file, }' --no-noop
Notice: Compiled catalog for managua.lan in environment production in 0.01 seconds
Notice: /Stage[main]/Main/File[/tmp/omgtest]/ensure: created
Notice: Applied catalog in 0.01 seconds
~~~

And yes, it'd be much better to use the default `--test`.

Reply:
Seems to work just fine with `puppet apply`. Last argument takes precedence:

~~~
$ puppet apply --no-noop -e 'file { "/tmp/omgtest": ensure => file, }' --noop
Notice: Compiled catalog for managua.lan in environment production in 0.01 seconds
Notice: /Stage[main]/Main/File[/tmp/omgtest]/ensure: current_value 'absent', should be 'file' (noop)
Notice: Class[Main]: Would have triggered 'refresh' from 1 event
Notice: Stage[main]: Would have triggered 'refresh' from 1 event
Notice: Applied catalog in 0.01 seconds
~~~

Reply:
I went for `puppet agent -t --splay` and the run *didn't* wait a random amount of time.

Reply:
@runejuhl I've added a `puppet_settings` variable.  It's an optional hash where you can override the default settings.